File: src/go/2_CreationalPatterns/1_Builder/builderfacets.go
package main

import "fmt"

type Person struct {
  StreetAddress, Postcode, City string
  CompanyName, Position string
  AnnualIncome int
}

type PersonBuilder struct {
  person *Person // needs to be inited
}

func NewPersonBuilder() *PersonBuilder {
  return &PersonBuilder{&Person{}}
}

func (it *PersonBuilder) Build() *Person {
  return it.person
}

func (it *PersonBuilder) Works() *PersonJobBuilder {
  return &PersonJobBuilder{*it}
}

func (it *PersonBuilder) Lives() *PersonAddressBuilder {
  return &PersonAddressBuilder{*it}
}

type PersonJobBuilder struct {
  PersonBuilder
}

func (pjb *PersonJobBuilder) At(
  companyName string) *PersonJobBuilder {
  pjb.person.CompanyName = companyName
  return pjb
}

func (pjb *PersonJobBuilder) AsA(
  position string) *PersonJobBuilder {
  pjb.person.Position = position
  return pjb
}

func (pjb *PersonJobBuilder) Earning(
  annualIncome int) *PersonJobBuilder {
  pjb.person.AnnualIncome = annualIncome
  return pjb
}

type PersonAddressBuilder struct {
  PersonBuilder
}

func (it *PersonAddressBuilder) At(
  streetAddress string) *PersonAddressBuilder {
  it.person.StreetAddress = streetAddress
  return it
}

func (it *PersonAddressBuilder) In(
  city string) *PersonAddressBuilder {
  it.person.City = city
  return it
}

func (it *PersonAddressBuilder) WithPostcode(
  postcode string) *PersonAddressBuilder {
  it.person.Postcode = postcode
  return it
}

func main() {
  pb := NewPersonBuilder()
  pb.
    Lives().
      At("123 London Road").
      In("London").
      WithPostcode("SW12BC").
    Works().
      At("Fabrikam").
      AsA("Programmer").
      Earning(123000)
  person := pb.Build()
  fmt.Println(*person)
}

File: src/swift/2_CreationalPatterns/1_Builder/BuilderFacets.swift
class Person : CustomStringConvertible
{
  // address
  var streetAddress = "", postcode = "", city = ""

  // employment
  var companyName = "", position = ""
  var annualIncome = 0

  var description: String {
    return "I live at \(streetAddress), \(postcode), \(city). " +
      "I work at \(companyName) as a \(position) earning \(annualIncome)."
  }
}

class PersonBuilder
{
  var person = Person()
  var lives : PersonAddressBuilder {
    return PersonAddressBuilder(person)
  }
  var works : PersonJobBuilder {
    return PersonJobBuilder(person)
  }

  // no implicit conversions, so...
  func build() -> Person
  {
    return person
  }
}

class PersonJobBuilder : PersonBuilder
{
  internal init(_ person:Person)
  {
    super.init()
    self.person = person
  }
  func at(_ companyName: String) -> PersonJobBuilder
  {
    person.companyName = companyName
    return self 
  }
  func asA(_ position: String) -> PersonJobBuilder
  {
    person.position = position
    return self
  }
  func earning(_ annualIncome: Int) -> PersonJobBuilder
  {
    person.annualIncome = annualIncome
    return self
  }
}

class PersonAddressBuilder : PersonBuilder
{
  internal init(_ person: Person)
  {
    super.init()
    self.person = person
  }
  func at(_ streetAddress: String) -> PersonAddressBuilder
  {
    person.streetAddress = streetAddress
    return self
  }
  func withPostcode(_ postcode: String) -> PersonAddressBuilder
  {
    person.postcode = postcode
    return self
  }
  func inCity(_ city: String) -> PersonAddressBuilder
  {
    person.city = city
    return self
  }
}

func main()
{
  let pb = PersonBuilder()
  let p = pb
    .lives.at("123 London Road").inCity("London").withPostcode("SW12BC")
    .works.at("Fabrikam").asA("Engineer").earning(123000)
    .build();
  print(p) 
}

main()
